test/sfs.py

```diff
@@ -1,28 +1,87 @@
 import ctypes
+import io
 
-c_helper = ctypes.CDLL('./pysfs.so')
+c_helper = ctypes.CDLL('../nfsv4/libnfs4.so')
 
-c_helper.create_client_py.argtypes = [ctypes.c_char_p]
-c_helper.open_file_py.argtypes = [ctypes.c_char_p, ctypes.c_char_p]
-c_helper.open_file_py.restype = ctypes.py_object
-c_helper.read_file_py.argtypes = [ctypes.py_object, ctypes.c_int, ctypes.c_int]
-c_helper.read_file_py.restype = ctypes.py_object
-c_helper.write_file_py.argtypes = [ctypes.py_object, ctypes.c_void_p, ctypes.c_int, ctypes.c_int]
+NFS4_RDONLY = 1
+NFS4_WRONLY = 2
+NFS4_RDWRITE = 3
+NFS4_CREAT = 4
+NFS4_TRUNC = 16
+
+NFS4_OK = 0
+NFS4_ENOENT = -2
+NFS4ERR_OPENMODE = 10038
+
+class Nfs4_struct(ctypes.Structure):
+    pass
+Nfs4 = ctypes.POINTER(Nfs4_struct) 
+
+class Nfs4_properties_struct(ctypes.Structure):
+    _fields_ = [
+        ('mask', ctypes.c_ulonglong),
+        ('name', ctypes.c_char * 256),
+        ('ino', ctypes.c_char * 128),
+        ('mode', ctypes.c_int),
+        ('nlink', ctypes.c_int),
+        ('user', ctypes.c_uint),
+        ('group', ctypes.c_uint),
+        ('size', ctypes.c_ulonglong),
+        ('type', ctypes.c_int),
+        ('access_time', ctypes.c_ulonglong),
+        ('modify_time', ctypes.c_ulonglong)
+    ]
+Nfs4_properties = ctypes.POINTER(Nfs4_properties_struct)
+
+class Nfs4_file_struct(ctypes.Structure):
+    pass
+Nfs4_file = ctypes.POINTER(Nfs4_file_struct)
+
+c_helper.nfs4_create.argtypes = [ctypes.c_char_p, ctypes.POINTER(Nfs4)]
+c_helper.nfs4_error_string.argtypes = [Nfs4]
+c_helper.nfs4_error_string.restype = ctypes.c_char_p
+c_helper.nfs4_open.argtypes = [Nfs4, ctypes.c_char_p, ctypes.c_int, Nfs4_properties, ctypes.POINTER(Nfs4_file)]
+c_helper.nfs4_pread.argtypes = [Nfs4_file, ctypes.c_void_p, ctypes.c_ulonglong, ctypes.c_ulonglong]
+c_helper.nfs4_pwrite.argtypes = [Nfs4_file, ctypes.c_void_p, ctypes.c_ulonglong, ctypes.c_ulonglong]
+
+client = Nfs4()
 
 def mount(host_ip):
     b_host_ip = host_ip.encode('utf-8')
-    c_helper.create_client_py(b_host_ip)
-
+    print("host_ip: " + host_ip)
+    print("create client")
+    if c_helper.nfs4_create(b_host_ip, ctypes.pointer(client)) != 0:
+        print("open client fail " + c_helper.nfs4_error_string(client))
+    
 def open(file_name, mode='r'):
-    fp = c_helper.open_file_py(file_name.encode('utf-8'), mode.encode('utf-8'))
-    return FileObjectWrapper(fp)
-
-def read(size):
-    c_helper.read_file_py(size)
+    #TODO: needs to validate MODE
+    p = Nfs4_properties_struct()
+    p.mask = 1<<33
+    p.mode = 0o666
+    
+    flags = 0
+    for c in mode:
+        if c == 'r':
+            flags |= NFS4_RDONLY
+        elif c == 'w':
+            flags |= (NFS4_TRUNC | NFS4_WRONLY | NFS4_CREAT)
+        elif c == 'a':
+            flags |= (NFS4_WRONLY | NFS4_CREAT)
```

Comment:
let's add an else error condition, raise `NotImplementedError` if we're still working on it.

Reply:
resolved